File: 817-DesignHashmap/817-DesignHashmap.java
// Last updated: 8/5/2025, 2:57:03 PM
class MyHashMap {

    class ListNode {
        int key, val;
        ListNode next;
        ListNode(int key, int val) {
            this.key = key;
            this.val = val;
        }
    }
    
    final ListNode[] nodes;
    
    private  int generateIdx(int key) {
        return Integer.hashCode(key) %  nodes.length;
    }
    
    private ListNode find(ListNode bucket, int key)  {
        ListNode node = bucket;
        ListNode prev = null;
        while(node!=null && node.key!=key) {
            prev = node;
            node = node.next;
        }
        return prev;
    }
    
    /** Initialize your data structure here. */
    public MyHashMap() {
        nodes = new ListNode[10000];
    }
    
    /** value will always be non-negative. */
    public void put(int key, int value) {
        int i  = generateIdx(key);
        if(nodes[i] == null)
            nodes[i] = new ListNode(-1,-1);
        ListNode prev = find(nodes[i],key);
        if(prev.next == null)
            prev.next = new ListNode(key,value);
        else
            prev.next.val = value;
    }
    
    /** Returns the value to which the specified key is mapped, or -1 if this map contains no mapping for the key */
    public int get(int key) {
        int i = generateIdx(key);
        if(nodes[i] == null)
            return -1;
        ListNode node = find(nodes[i],key);
        return node.next == null ? -1 : node.next.val;
    }
    
    /** Removes the mapping of the specified value key if this map contains a mapping for the key */
    public void remove(int key) {
        int i = generateIdx(key);
        if(nodes[i] == null)
            return;
        ListNode prev = find(nodes[i],key);
        if(prev.next  == null)  
            return;
        else
            prev.next = prev.next.next;
    }
}

/**
 * Your MyHashMap object will be instantiated and called as such:
 * MyHashMap obj = new MyHashMap();
 * obj.put(key,value);
 * int param_2 = obj.get(key);
 * obj.remove(key);
 */

File: 817-DesignHashmap/817-DesignHashmap.py
# Last updated: 8/5/2025, 2:57:01 PM
class MyHashMap:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.hash = 2069
        self.buckets = []
        for ele in range(self.hash):
            self.buckets.append([])
        

    def put(self, key: int, value: int) -> None:
        """
        value will always be non-negative.
        """
        bucket = self.buckets[key % self.hash]
        item = next((item for item in bucket if item[0] == key), None)
        if item is not None:
            item[1] = value
        else:
            bucket.append([key, value])

    def get(self, key: int) -> int:
        """
        Returns the value to which the specified key is mapped, or -1 if this map contains no mapping for the key
        """
        bucket = self.buckets[key % self.hash]
        item = next((item for item in bucket if item[0] == key), None)
        if item is not None:
            return item[1]
        else:
            return -1

    def remove(self, key: int) -> None:
        """
        Removes the mapping of the specified value key if this map contains a mapping for the key
        """
        bucket = self.buckets[key % self.hash]
        item = next((item for item in bucket if item[0] == key), None)
        if item is not None:
            bucket.remove(item)


# Your MyHashMap object will be instantiated and called as such:
# obj = MyHashMap()
# obj.put(key,value)
# param_2 = obj.get(key)
# obj.remove(key)